x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/TokenPruningConfig.java

```diff
@@ -0,0 +1,162 @@
+        this.onlyScorePrunedTokens = in.readBoolean();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeFloat(tokensFreqRatioThreshold);
+        out.writeFloat(tokensWeightThreshold);
+        out.writeBoolean(onlyScorePrunedTokens);
+    }
+
+    /**
+     * Returns the frequency ratio threshold to apply on the query.
+     * Tokens whose frequency is more than ratio_threshold times the average frequency of all tokens in the specified
+     * field are considered outliers and may be subject to removal from the query.
+     */
+    public float getTokensFreqRatioThreshold() {
+        return tokensFreqRatioThreshold;
+    }
+
+    /**
+     * Returns the weight threshold to apply on the query.
+     * Tokens whose weight is more than (weightThreshold * best_weight) of the highest weight in the query are not
+     * considered outliers, even if their frequency exceeds the specified ratio_threshold.
+     * This threshold ensures that important tokens, as indicated by their weight, are retained in the query.
+     */
+    public float getTokensWeightThreshold() {
+        return tokensWeightThreshold;
+    }
+
+    /**
+     * Returns whether the filtering process retains tokens identified as non-relevant based on the specified thresholds
+     * (ratio and weight). When {@code true}, only non-relevant tokens are considered for matching and scoring documents.
+     * Enabling this option is valuable for re-scoring top hits retrieved from a {@link WeightedTokensQueryBuilder} with
+     * active thresholds.
+     */
+    public boolean isOnlyScorePrunedTokens() {
+        return onlyScorePrunedTokens;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        TokenPruningConfig that = (TokenPruningConfig) o;
+        return Float.compare(that.tokensFreqRatioThreshold, tokensFreqRatioThreshold) == 0
+            && Float.compare(that.tokensWeightThreshold, tokensWeightThreshold) == 0
+            && onlyScorePrunedTokens == that.onlyScorePrunedTokens;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(tokensFreqRatioThreshold, tokensWeightThreshold, onlyScorePrunedTokens);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(PRUNING_CONFIG.getPreferredName());
+        builder.field(TOKENS_FREQ_RATIO_THRESHOLD.getPreferredName(), tokensFreqRatioThreshold);
+        builder.field(TOKENS_WEIGHT_THRESHOLD.getPreferredName(), tokensWeightThreshold);
+        if (onlyScorePrunedTokens) {
+            builder.field(ONLY_SCORE_PRUNED_TOKENS_FIELD.getPreferredName(), onlyScorePrunedTokens);
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    public static TokenPruningConfig fromXContent(XContentParser parser) throws IOException {
```

Comment:
IMO, this type of fromXContent parsing is very opaque and bug inducing. Just because our query builders are nasty doesn't mean a simple pojo like this should be :).

Take a look here: https://github.com/elastic/elasticsearch/blob/8e6e0e59ea34be29a6495f0680944d6a7f559756/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/AnalysisConfig.java#L52

This shows how to use a static parser with parse fields. Its way easier to reason about and you don't have all this weird while loop and else if things.

Reply:
Don't worry about "lenient" vs. "strict" here. My guess is that we are always "strict" and don't allow unknown fields to be input.

Reply:
Thanks for the pointer. I agree we could make this better, but I'd prefer to classify this as something that can happen later, in a future PR.